src/bin/cargo-miri.rs

```diff
@@ -301,34 +301,19 @@ fn setup(ask_user: bool) {
                 .stdout;
             let sysroot = std::str::from_utf8(&sysroot).unwrap();
             let sysroot = Path::new(sysroot.trim_end_matches('\n'));
-            // First try: `$SYSROOT/lib/rustlib/src/rust`; test if that contains `Cargo.lock`.
-            let rustup_src = sysroot.join("lib").join("rustlib").join("src").join("rust");
-            let base_dir = if rustup_src.join("Cargo.lock").exists() {
-                // Just use this.
-                rustup_src
-            } else {
-                // Maybe this is a local toolchain built with `x.py` and linked into `rustup`?
-                // Second try: `$SYSROOT/../../..`; test if that contains `x.py`.
-                let local_src = sysroot.parent().and_then(Path::parent).and_then(Path::parent);
-                match local_src {
-                    Some(local_src) if local_src.join("x.py").exists() => {
-                        // Use this.
-                        PathBuf::from(local_src)
-                    }
-                    _ => {
-                        // Fallback: Ask the user to install the `rust-src` component, and use that.
-                        let mut cmd = Command::new("rustup");
-                        cmd.args(&["component", "add", "rust-src"]);
-                        ask_to_run(
-                            cmd,
-                            ask_user,
-                            "install the rustc-src component for the selected toolchain",
-                        );
-                        rustup_src
-                    }
-                }
-            };
-            base_dir.join("src") // Xargo wants the src-subdir
+            // Check for `$SYSROOT/lib/rustlib/src/rust/src`; test if that contains `libstd/lib.rs`.
+            let rustup_src = sysroot.join("lib").join("rustlib").join("src").join("rust").join("src");
```

Comment:
Hmm I should note something: `$sysroot/lib/rustlib/src/rust` corresponds to the "root" of `rust-lang/rust`, that's why it has a further `src` in it. It's annoying that it repeats but I think it would be to work with if `$sysroot/lib/rustlib/src/rust` was the directory being referred to and `src/libstd/lib.rs` a path within that directory.

Sadly it seems like `XARGO_RUST_SRC` includes the `src` directory component of `rust-lang/rust`, so this entire comment is unactionable :disappointed:.

Reply:
Yeah it would be nicer to treat `rustlib/src/rust` as the root but as you said, that's not what xargo did, and now it's too late to change it (and not worth trying anyway).